trieste/acquisition/optimizer.py

```diff
@@ -197,68 +199,172 @@ def optimize_continuous(
+        gr.switch(starting_points[i].numpy(), bounds[i], optimizer_args)
+        for i, gr in enumerate(child_greenlets)
+    ]
+
+    while True:
+        all_done = True
+        for i, result in enumerate(child_results):  # Process results from children.
+            if isinstance(result, spo.OptimizeResult):
+                continue  # children return a `spo.OptimizeResult` if they are finished
+            all_done = False
+            assert isinstance(result, np.ndarray)  # or an `np.ndarray` with the query `x` otherwise
+            np_batch_x[i, :] = result
+
+        if all_done:
+            break
+
+        # Batch evaluate query `x`s from all children.
+        batch_x = tf.constant(np_batch_x, dtype=tf_dtype)  # [num_optimization_runs, d]
+        batch_y, batch_dy_dx = _objective_value_and_gradient(batch_x)
+        np_batch_y = batch_y.numpy().astype("float64")
+        np_batch_dy_dx = batch_dy_dx.numpy().astype("float64")
+
+        for i, greenlet in enumerate(child_greenlets):  # Feed `y` and `dy_dx` back to children.
+            if greenlet.dead:  # Allow for crashed greenlets
+                continue
+            child_results[i] = greenlet.switch(np_batch_y[i], np_batch_dy_dx[i, :])
+
+    best_run_id = np.argmax([-result.fun for result in child_results])  # identify best solution
+    np_chosen_point = child_results[best_run_id].x.reshape(1, -1)  # [1, D]
+    success = np.any(
+        [result.success for result in child_results]
+    )  # Check that at least one optimization was successful
+
+    return tf.constant(np_chosen_point, dtype=tf_dtype), success  # convert back to TF
+
+
+class ScipyLbfgsBGreenlet(gr.greenlet):
+    """
+    Worker greenlet that runs a single Scipy L-BFGS-B. Each greenlet performs all the L-BFGS-B
+    update steps required for an individual optimization. However, the evaluation
+    of our acquisition function (and its gradients) is delegated back to the main Tensorflow
+    process (the parent greenlet) where evaluations can be made efficiently in parallel.
+    """
+
+    def run(
+        self, start: np.ndarray, bounds: spo.Bounds, optimizer_args: dict[str, Any] = dict()
+    ) -> spo.OptimizeResult:
+        cache_x = start + 1  # Any value different from `start`.
+        cache_y: Optional[np.ndarray] = None
+        cache_dy_dx: Optional[np.ndarray] = None
+
+        def value_and_gradient(
+            x: np.ndarray,
+        ) -> Tuple[np.ndarray, np.ndarray]:  # Collect function evaluations from parent greenlet
+            nonlocal cache_x
+            nonlocal cache_y
+            nonlocal cache_dy_dx
+
+            if not (cache_x == x).all():
+                cache_x[:] = x  # Copy the value of `x`. DO NOT copy the reference.
+                # Send `x` to parent greenlet, which will evaluate all `x`s in a batch.
+                cache_y, cache_dy_dx = self.parent.switch(cache_x)
+
+            return cache_y, cache_dy_dx
+
+        return spo.minimize(
```

Comment:
how come we are switching to Scipy directly from GPflow's wrapper?
(or why were we using GPflow's wrapper in the first place)

Reply:
we have essentially made our own wrapper that can do batching.